pkg/gatewayserver/io/basicstationlns/basicstationlns_test.go

```diff
@@ -758,3 +758,261 @@ func TestTraffic(t *testing.T) {
 	}
 
 }
+
+var MuxTime, RxTime float64
+
+func TestRTT(t *testing.T) {
+	a := assertions.New(t)
+	ctx := log.NewContext(test.Context(), test.GetLogger(t))
+
+	c := component.MustNew(test.GetLogger(t), &component.Config{
+		ServiceBase: config.ServiceBase{
+			HTTP: config.HTTP{
+				Listen: ":8100",
+			},
+		},
+	})
+	gs := mock.NewServer()
+	srv := New(ctx, gs)
+	c.RegisterWeb(srv)
+	test.Must(nil, c.Start())
+	defer c.Close()
+
+	gs.RegisterGateway(ctx, registeredGatewayID, &registeredGateway)
+
+	wsConn, _, err := websocket.DefaultDialer.Dial(testTrafficEndPoint, nil)
+	if !a.So(err, should.BeNil) {
+		t.Fatalf("Connection failed: %v", err)
+	}
+	defer wsConn.Close()
+
+	var gsConn *io.Connection
+	select {
+	case gsConn = <-gs.Connections():
+	case <-time.After(timeout):
+		t.Fatal("Connection timeout")
+	}
+
+	for _, tc := range []struct {
+		Name                   string
+		InputBSUpstream        interface{}
+		InputNetworkDownstream *ttnpb.DownlinkMessage
+		InputDownlinkPath      *ttnpb.DownlinkPath
+		WaitTime               time.Duration
+		ExpectedRTTStatsCount  int
+	}{
+		{
+			Name: "JoinRequest",
+			InputBSUpstream: messages.JoinRequest{
+				MHdr:     0,
+				DevEUI:   basicstation.EUI{Prefix: "DevEui", EUI64: types.EUI64{0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11}},
+				JoinEUI:  basicstation.EUI{Prefix: "JoinEui", EUI64: types.EUI64{0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22}},
+				DevNonce: 18000,
+				MIC:      12345678,
+				RadioMetaData: messages.RadioMetaData{
+					DataRate:  1,
+					Frequency: 868300000,
+					UpInfo: messages.UpInfo{
+						RxTime: 1548059982,
+						XTime:  12666373963464220,
+						RSSI:   89,
+						SNR:    9.25,
+					},
+				},
+			},
+			ExpectedRTTStatsCount: 0,
+		},
+		{
+			Name: "Downlink",
+			InputNetworkDownstream: &ttnpb.DownlinkMessage{
+				RawPayload: []byte("Ymxhamthc25kJ3M=="),
+				EndDeviceIDs: &ttnpb.EndDeviceIdentifiers{
+					DeviceID: "testdevice",
+					DevEUI:   eui64Ptr(types.EUI64{0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11}),
+				},
+				Settings: &ttnpb.DownlinkMessage_Request{
+					Request: &ttnpb.TxRequest{
+						Class:            ttnpb.CLASS_A,
+						Priority:         ttnpb.TxSchedulePriority_NORMAL,
+						Rx1Delay:         ttnpb.RX_DELAY_1,
+						Rx1DataRateIndex: 5,
+						Rx1Frequency:     868100000,
+					},
+				},
+				CorrelationIDs: []string{"correlation1", "correlation2"},
+			},
+
+			InputDownlinkPath: &ttnpb.DownlinkPath{
+				Path: &ttnpb.DownlinkPath_UplinkToken{
+					UplinkToken: io.MustUplinkToken(ttnpb.GatewayAntennaIdentifiers{GatewayIdentifiers: registeredGatewayID}, 1553759666),
+				},
+			},
+		},
+		{
+			Name: "FollowUpTxAck",
+			InputBSUpstream: messages.TxConfirmation{
+				Diid:  1,
+				XTime: 1548059982,
+			},
+			ExpectedRTTStatsCount: 1,
+			WaitTime:              2 * test.Delay,
```

Comment:
Sleep longer; `test.Delay` is 1ms without test slow down. Safer is to use `1 << 3` and higher